onmt/utils/ExtendedCmdLine.lua

```diff
@@ -0,0 +1,325 @@
+-- Local utility functions
+------------------------------------------------------------------------------------------------------------------
+
+--[[ Convert `val` string to its actual type (boolean, number or string). ]]
+local function convert(key, val, ref)
+  local new
+
+  if type(val) == type(ref) then
+    new = val
+  elseif type(ref) == 'boolean' then
+    if val == 'true' then
+      new = true
+    elseif val == 'false' then
+      new = false
+    else
+      error('option ' .. key .. ' expects a boolean value [true, false]')
+    end
+  elseif type(ref) == 'number' then
+    new = tonumber(val)
+    assert(new ~= nil, 'option ' .. key .. ' expects a number value')
+  end
+
+  return new
+end
+
+------------------------------------------------------------------------------------------------------------------
+
+local extendedCmdLine, parent, path
+-- the utils function can be run without torch
+if torch then
+  extendedCmdLine, parent = torch.class('onmt.extendedCmdLine', 'torch.CmdLine')
+  path = require('pl.path')
+else
+  extendedCmdLine = {}
+end
+
+--[[
+  Extended handling of command line options - provide validation methods, and utilities for handling options
+  at module level.
+
+  For end-user - provides '-h' for help, with possible '-md' variant for preparing MD ready help file.
+
+  Provides also possibility to read options from config file with '-config file' or save current options to
+  config file with '-save_config file'
+
+  Example:
+
+    cmd = onmt.utils.ExtendedCmdLine.new()
+    local optim_options = {
+      {'-max_batch_size',     64   , 'Maximum batch size', {valid=onmt.utils.ExtendedCmdLine.isUInt()}}
+    }
+
+    cmd:setCmdLineOptions(optim_options, 'Optimization')
+    local opt = cmd:parse(arg)
+
+    local optimArgs = cmd.getModuleOpts(opt, optim_options)
+]]
+
+function extendedCmdLine:__init(script)
+  self.script = script
+  parent.__init(self)
+end
+
+function extendedCmdLine:help(arg, doMd)
+  if doMd then
+    for _,option in ipairs(self.helplines) do
+      if type(option) == 'table' then
+        io.write('* ')
+        if option.default ~= nil then -- it is an option
+          io.write("`"..option.key.."`: ")
+          if option.meta and option.meta.enum then
+            io.write(' ('..table.concat(option.meta.enum, ', ')..') ')
+          end
+          option.help = option.help:gsub(" *\n   *"," ")
+          if option.help then io.write(option.help) end
+          io.write(' [' .. tostring(option.default) .. ']')
+        else -- it is an argument
+          io.write('<' .. onmt.utils.String.stripHyphens(option.key) .. '>')
+          if option.help then io.write(' ' .. option.help) end
+        end
+      else
+        local display = option:gsub("%*","-")
+        io.write(display) -- just some additional help
+      end
+      io.write('\n')
+    end
+    io.write('\n')
+    io.write('* `-h`: this help file.\n')
+    io.write('* `-config`: read options from config file.\n')
```

Comment:
Why not call `setCmdLineOptions` in the constructor for `-config` and `-save_config`. It would avoid manually handling them in `help()` and `parse()`.